rfcs/convergence/simplify-prop-merging.md

```diff
@@ -0,0 +1,359 @@
+
+#### Current usage of `mergeProps()`
+
+```tsx
+const state = mergeProps(
+  {
+    ref: resolvedRef,
+    as: 'button',
+    icon: { as: 'span' },
+    content: { as: 'span', children: props.children },
+    loader: { as: Loader, role: 'presentation' },
+  },
+  defaultProps,
+  resolveShorthandProps(props, buttonShorthandProps),
+);
+```
+
+#### Proposed usage
+
+```tsx
+const state: FooState = {
+  as: "button",
+
+  components: {
+    icon: "span",
+    content: "span",
+    loader: Loader,
+  },
+
+  // user props
+  ...props,
+
+  ref: resolvedRef,
+  content: resolveShorthand(
+    props.content,
+    { children: props.children } /* defaultProps for a slot */
+  ),
+  icon: resolveShorthand(props.icon),
+  loader: resolveShorthand(
+    props.loader,
+    { role: "presentation" } /* defaultProps for a slot */
+  ),
+};
+
+export interface FooProps extends ComponentProps, React.ButtonHTMLAttributes<HTMLButtonElement> {
+  icon?: ShorthandProps<React.HTMLAttributes<HTMLSpanElement>;
+  content?: ShorthandProps<React.HTMLAttributes<HTMLSpanElement>;
+  Loader?: ShorthandProps<LoaderProps>;
+};
+
+export interface FooState extends FooProps, React.ButtonHTMLAttributes<HTMLButtonElement> {
+  icon: ObjectShorthandProps<React.HTMLAttributes<HTMLSpanElement>;
+  content: ObjectShorthandProps<React.HTMLAttributes<HTMLSpanElement>;
+  Loader: ObjectShorthandProps<LoaderProps>;
+};
+```
+
+Typings there are 100% clear since they directly map to types 🎉
+
+Now when either internally or externally we want to extend/add slots:
+
+```ts
+const state = useComponentState();
+state.components.extraSlot = 'div';
+state.extraSlot = resolveShorthand({
+  /** Shorthand props */
+});
+
+state.components.icon = 'button';
+state.icon = filterNativeButtonProps(state.icon);
+```
+
+This results in the use of standard object spread and explicit resolution of slots in state, rather than deep merging all user and default props.
+
+Without `defaultProps` there is now only one way to extend component state - by mutating it.
+
+Mutating state has **always** been possible, and can only be stopped if we decide to introduce true immutability.
+
+##### Basic implementation of `resolveShorthand`
+
+```tsx
+function resolveShorthand<T extends Record<string, any>>(
+  value: ShorthandProps<T>,
+  defaultProps?: T,
+): ObjectShorthandProps<T> {
+  let resolvedShorthand: ObjectShorthandProps<T> = {} as T;
+
+  if (typeof value === 'string' || typeof value === 'number' || React.isValidElement(value)) {
+    resolvedShorthand = { children: value } as ObjectShorthandProps<T>;
+  }
+
+  if (typeof value === 'object') {
+    resolvedShorthand = value as ObjectShorthandProps<T>;
+  }
+
+  return { ...(defaultProps as ObjectShorthandProps<T>), ...resolvedShorthand };
```

Comment:
so much casting makes me feel  uneasy. can we note that this is a `contrived/simplified` implementation example